### PR TITLE
Add a test demonstrating connect timeout functionality

### PR DESCRIPTION
**Pull request by romange**

### CODE REVIEW

File: test/connect.c [hunk 1]
```diff
@@ -105,13 +105,9 @@ static int listen_on_socket(int fd)
...
 }
 
-static int connect_socket(struct io_uring *ring, int fd, int *code)
-{
-	struct io_uring_sqe *sqe;
-	struct sockaddr_in addr;
-	int ret, res, val = 1;
-	socklen_t code_len = sizeof(*code);
-
+static int configure_connect(int fd, struct sockaddr_in* addr) {
```
Comment (isilence) [inline test/connect.c:108]: brackets for function body should be on next line

File: test/connect.c [hunk 2]
```diff
@@ -235,6 +244,86 @@ static int test_connect(struct io_uring *ring)
...
+	io_uring_prep_connect(sqe, connect_fd, (struct sockaddr*)&addr, sizeof(addr));
+	sqe->user_data = 1;
+	sqe->flags |= IOSQE_IO_LINK;
+
+	sqe = io_uring_get_sqe(ring);
+	if (!sqe) {
+		fprintf(stderr, "unable to get sqe\n");
+		goto err;
+	}
+	struct __kernel_timespec ts;
```
Comment (isilence) [inline test/connect.c:304]: please, keep var declarations in the beginning of a body section.

File: test/connect.c [hunk 3]
```diff
@@ -235,6 +244,86 @@ static int test_connect(struct io_uring *ring)
...
+	}
+
+	// Fill up available place in the queue
+	for (int i = 0; i < 1; ++i) {
+		fd[i] = create_socket();
+		if (configure_connect(connect_fd, &addr) == -1)
+			goto err;
+		if (connect(fd[i], &addr, sizeof(addr)) == -1) {
+			fprintf(stderr, "unable to connect %d\n", errno);
+			goto err;
```
Comment (isilence) [inline test/connect.c:285]: A trailing tab after err;

File: test/connect.c [hunk 4]
```diff
@@ -235,6 +244,86 @@ static int test_connect(struct io_uring *ring)
...
+	sqe->flags |= IOSQE_IO_LINK;
+
+	sqe = io_uring_get_sqe(ring);
+	if (!sqe) {
+		fprintf(stderr, "unable to get sqe\n");
+		goto err;
+	}
+	struct __kernel_timespec ts;
+	ts.tv_sec = 0;
+	ts.tv_nsec = 100000;
```
Comment (isilence) [inline test/connect.c:306]: `cqe->user_data = 2;`
for code in the comment below to work

File: test/connect.c [hunk 5]
```diff
@@ -235,6 +244,86 @@ static int test_connect(struct io_uring *ring)
...
+	
+	io_uring_prep_link_timeout(sqe, &ts, 0);
+	ret = submit_and_wait(ring, 2, &res);
```
Comment (isilence) [inline test/connect.c:309]: It doesn't work this way, linked timeouts doesn't return an error directly but rather generate an CQE,
so you will get 2 CQEs. Tried it out a bit rewritten, as expected there are 2 CQEs, -ECANCELLED for
the connect and -ETIME for the timeout.

E.g. should be as this

```C
        ...
	ret = io_uring_submit(ring);
	if (ret != 2) {
		fprintf(stderr, "submitted %d\n", ret);
		return 1;
	}

	for (i = 0; i < 2; i++) {
		int expected;

		ret = io_uring_wait_cqe(ring, &cqe);
		if (ret) {
			fprintf(stderr, "wait_cqe=%d\n", ret);
			return 1;
		}

		expected = (cqe->user_data == 1) ? -ECANCELED : -ETIME;
		if (expected != cqe->res) {
			fprintf(stderr, "cqe %d, res %d, wanted %d\n",
				(int)cqe->user_data, cqe->res, expected);
			goto err;
		}
		io_uring_cqe_seen(ring, cqe);
	}
        ...
```

Comment (romange) [reply test/connect.c]: Thanks!

**Comment by romange** [comment]
please take a look, it passed now.

File: test/connect.c [hunk 6]
```diff
@@ -235,6 +245,101 @@ static int test_connect(struct io_uring *ring)
...
+
+		ret = io_uring_wait_cqe(ring, &cqe);
+		if (ret) {
+			fprintf(stderr, "wait_cqe=%d\n", ret);
+			return 1;
+		}
+
+		expected = (cqe->user_data == 1) ? -ECANCELED : -ETIME;
+		if (expected != cqe->res) {
+			fprintf(stderr, "cqe %d, res %d, wanted %d\n", (int)cqe->user_data, cqe->res, expected);
```
Comment (isilence) [inline test/connect.c:322]: Don't remember whether Jens tries to keep width to 80

File: test/connect.c [hunk 7]
```diff
@@ -235,6 +245,101 @@ static int test_connect(struct io_uring *ring)
 	return -1;
 }
 
+static int test_connect_timeout(struct io_uring *ring)
+{
+	int fd = -1, connect_fd = -1, accept_fd = -1;
+	int re;
```
Comment (isilence) [inline test/connect.c:251]: s/re/ret/
```
     CC connect
connect.c: In function ‘test_connect_timeout’:
connect.c:267:2: error: ‘ret’ undeclared (first use in this function); did you mean ‘re’?
  267 |  ret = bind(accept_fd, (struct sockaddr*)&addr, sizeof(addr));
      |  ^~~
      |  re
connect.c:267:2: note: each undeclared identifier is reported only once for each function it appears in
connect.c:251:6: warning: unused variable ‘re’ [-Wunused-variable]
  251 |  int re;
      |      ^~
```

File: test/connect.c [hunk 8]
```diff
@@ -263,6 +368,12 @@ int main(int argc, char *argv[])
 		return 1;
 	}
 
+	ret = test_connect_timeout(&ring);
+	if (ret == -1) {
```
Comment (isilence) [inline test/connect.c:372]: Let's make it `if (ret)`, because almost all tests here return 0 - success, 1 - failure.

**Review by isilence — COMMENTED** [review]
If with ret change, looks good to me.

File: test/connect.c (hunk 8 above)
Comment (romange) [reply test/connect.c:372]: I am actually being consistent with all the other tests here.

**Comment by romange** [comment]
Fixed.

File: test/connect.c (hunk 8 above)
Comment (isilence) [reply test/connect.c:373]: Frankly, I don't really care whether -1 or 1, it's just often 1 around tests. If you think `ret == -1` is better, I don't mind, just please make sure  `test_connect_timeout()` doesn't return 1 (as in hunks I suggested).

**Comment by isilence** [comment]
> Fixed.

Thanks a lot!

File: test/connect.c (hunk 8 above)
Comment (romange) [reply test/connect.c:373]: done

File: test/connect.c [hunk 9]
```diff
@@ -235,6 +245,102 @@ static int test_connect(struct io_uring *ring)
...
+	}
+
+	for (int i = 0; i < 2; i++) {
+		int expected;
+		struct io_uring_cqe *cqe;
+
+		ret = io_uring_wait_cqe(ring, &cqe);
+		if (ret) {
+			fprintf(stderr, "wait_cqe=%d\n", ret);
+			return 1;
```
Comment (isilence) [inline test/connect.c:317]: umm...
but I think it's good enough and we'll fix it on top. Thanks!

Comment (romange) [reply test/connect.c]: no, I am sorry. lack of attention from late hours.

**Comment by romange** [comment]
@axboe  Can you please take a look?

**Comment by isilence** [comment]
> @axboe Can you please take a look?

Looks good to me now
Thanks for finishing it! I know it may be annoying.

**Comment by axboe** [comment]
Thanks! LGTM too.

**Pull request merged**